### PR TITLE
add cache calculator

### LINKED ISSUE #1086: add kv cache calculator
### Describe the workflow you want to enable

Want to allow to query for the cache size without having to actually run the model. This helps theoretical analysis and also practically to prevent OOM errors.

### Describe your proposed solution

Add an API to estimate the size of the cache given a model.

### Describe alternatives you've considered, if relevant

_No response_

### Additional context

_No response_

### Impact

None

### PR DESCRIPTION
**Pull request by jmkuebler**
## Issue

This fixes issue PriorLabs/TabPFN#1086 <br>Fixes PriorLabs/TabPFN#1086

## Motivation and Context

When analyzing the size of the Cache it is good to have a theoretical model to query without having to really build the cache

---

## Public API Changes

- [X] No Public API changes
- [ ] Yes, Public API changes (Details below)

---

## How Has This Been Tested?

Added unit tests that compare the estimate with the actual cache size. Also for TabPFNv3 we test actual numbers we computed ourselves.

---

## Checklist

- [X] The changes have been tested locally.
- [ ] Documentation has been updated (if the public API or usage changes).
- [X] A changelog entry has been added (see `changelog/README.md`), or "no changelog needed" label requested.
- [X] The code follows the project's style guidelines.
- [X] I have considered the impact of these changes on the public API.

---

### CODE REVIEW

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 1]
```diff
@@ -251,6 +256,90 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+       them. Controlled by ``include_decoder_activations`` (on by default).
+
+    The KV cache and the activations are stored at *different* dtypes: the
+    default engine int8-quantizes the KV cache (``quantize_kv_cache``) but leaves
+    ``train_embeddings`` at the compute dtype. Each term is therefore sized with
+    its own dtype rather than a single shared one. The ``scaler_cache`` /
+    ``inducing_hidden`` entries are small and not counted.
+
+    Args:
+        X_train: Training features; only ``X_train.shape[0]`` (the row count)
```
Comment (ggprior) [inline src/tabpfn/architectures/tabpfn_v3.py]: expected only the shape would be relevant here, even better if just the row count -- no need to pass a tensor in that case, perhaps we'd just accept `n_rows` here or similar
(sorry my initial function definition was bad)

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 2]
```diff
@@ -251,6 +256,90 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+    *,
+    model_config: TabPFNV3Config,
+    ensemble_size: int,
+    dtype: torch.dtype,
+    quantize_kv_cache: bool = True,
+    include_decoder_activations: bool = True,
+) -> int:
+    """Estimate the cached memory in bytes for a TabPFN v3 inference run.
+
+    Analytical counterpart of the (removed) ``TabPFNV3Cache.cache_size_mb``:
```
Comment (ggprior) [inline src/tabpfn/architectures/tabpfn_v3.py]: minor: this is reminiscing about a past state of the code base, which tends to get confusing over time

**Review by arthur-priorlabs — APPROVED** [review]
Looks good, although this doesn't seem to be used anywhere in the codebase.

Maybe add some kind of early warning for `fit_with_cache` when available memory is too low to handle it?

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 3]
```diff
@@ -251,6 +256,90 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
         )
 
 
+def estimate_cache_size(
```
Comment (ggprior) [inline src/tabpfn/architectures/tabpfn_v3.py]: Can we say more here than `estimate_`? What are the conditions for this to be exact? If it's an estimation, how far off can we expect this to be from the true value?
(note: Ideal for the engineering use case would be an upper bound)

Comment (ggprior) [reply src/tabpfn/architectures/tabpfn_v3.py]: In fact we may be underselling ourselves here as the tests look like exact assertions!

**Review by ggprior — COMMENTED** [review]
This will be a useful feature, thanks for working on it! Some comments below, refer to @bejaeger for approvals

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 1 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: That is a good call -- at least for v3. But potentially future models could depend on columns again. Also for a user its maybe more naive to compute the cache using your dataset as input whether some derivative of it.

Just some thoughts, happy to change to number of rows if you think that's better, and so far the method is anyways for v3 specifically.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 3 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: yes the kv and final embeddings is accurate. But so far we are excluding the inducing points. So it is a slight "underestimate". But the inducing points don't scale with the number of rows.

Maybe it is best to include them as well and then it is an accurate "computation" of everything that is cached.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 1 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: I updated it and the method now also needs the column count, as it includes also the inducing points now. 

I overloaded the X_train argument so we can also just pass a tuple of `rows, columns`.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 3 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: I added all the other terms even if they are very small and refactored the method to `calculate`

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 1 above)
Comment (ggprior) [reply src/tabpfn/architectures/tabpfn_v3.py]: > But potentially future models could depend on columns again

good call, current approach of overloading also works 👍

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 3 above)
Comment (ggprior) [reply src/tabpfn/architectures/tabpfn_v3.py]: sounds good!

File: src/tabpfn/architectures/kv_cache.py [hunk 4]
```diff
@@ -19,6 +19,11 @@
 import torch
 from torch import Tensor
 
+# Default integer dtype the KV cache is quantized to. Single source of truth
+# shared by :meth:`KVCacheEntry.quantize` and any cache-size estimation, so the
+# stored dtype is never hardcoded in two places.
```
Comment (bejaeger) [inline src/tabpfn/architectures/kv_cache.py]: nit: comment seems unnecessarily long

File: src/tabpfn/architectures/kv_cache.py [hunk 5]
```diff
@@ -78,11 +83,13 @@ def to(self, device: torch.device | str) -> KVCacheEntry:
...
 
-    def quantize(self, dtype: torch.dtype = torch.int8) -> QuantizedKVCacheEntry:
+    def quantize(
+        self, dtype: torch.dtype = QUANTIZED_KV_DTYPE
+    ) -> QuantizedKVCacheEntry:
         """Quantize this entry with per-tensor symmetric scaling.
 
         Args:
-            dtype: Target integer dtype (default ``torch.int8``).
+            dtype: Target integer dtype (default :data:`QUANTIZED_KV_DTYPE`).
```
Comment (bejaeger) [inline src/tabpfn/architectures/kv_cache.py]: nit: Let's remove the `:data:` or `:method:` hints since we're not planning to generate docs from them.

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 6]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+       compress the train rows into a fixed set) and instead scales with the
+       column count, so it can dominate on wide, small-``N`` datasets. Controlled
+       by ``include_inducing_points`` (on by default).
+
+    4. The fitted scaler stats ``scaler_cache`` (``mean`` + ``std``), of shape
+       ``(2, n_features)`` per estimator. Always counted (small).
+
+    Each term is sized with its own dtype: the default engine int8-quantizes the
+    KV cache (``quantize_kv_cache``) but leaves everything else at the compute
+    ``dtype``.
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: the doc string seems overly long for a method that can be easily be read in code. Could we shorten and potentially move some of the high-level comments to the code.

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 7]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
         )
 
 
+def calculate_cache_size(
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: nit

```suggestion
def get_cache_size(
```

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 8]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+         (GQA), else ``icl_num_heads`` (plain MHA).
+
+       When ``quantize_kv_cache`` is True the K/V are int8 plus one scalar scale
+       per key and per value tensor (``nlayers * 2`` scalars at ``dtype``).
+
+    2. The last-layer train activations ``train_embeddings`` of shape
+       ``(N_train, embed_dim * feat_agg_num_cls_tokens)`` per estimator -- **only
+       for classification** (binary or multiclass), where the ``ManyClassDecoder``
+       attends test queries to them as retrieval keys. Regression never uses
+       them. Controlled by ``include_decoder_activations`` (on by default).
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: Currently, we're actually a bit wasteful and also include these activations for the cache for regression.
Also, I don't like the API here because the user has to think "do I need to include it or not". Could we just parse the task_type instead?
For now though, I think we can simply remove it until we change the cache storage for regression.

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 9]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
 
+def calculate_cache_size(
+    X_train: Any,
+    *,
+    model_config: TabPFNV3Config,
+    ensemble_size: int,
+    dtype: torch.dtype,
+    quantize_kv_cache: bool = True,
+    include_decoder_activations: bool = True,
+    include_inducing_points: bool = True,
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: What use case do you intend here?

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 10]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
         )
 
 
+def calculate_cache_size(
+    X_train: Any,
+    *,
+    model_config: TabPFNV3Config,
+    ensemble_size: int,
+    dtype: torch.dtype,
+    quantize_kv_cache: bool = True,
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py:265]: should we make this more explicit and have `kv_cache_dtype` directly? Maybe also change `dtype` to `base_dtype` in that case? 
By moving the responsibility to the function caller we don't have to change this fn when we change the quantization.

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 11]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
         )
 
 
+def calculate_cache_size(
+    X_train: Any,
+    *,
+    model_config: TabPFNV3Config,
+    ensemble_size: int,
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: nit: I'm not a fan of having this here, as the model itself shouldn't really care about the ensemble size. Could we remove and move the responsibility to the caller? WDYT?

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 12]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+    total_bytes = kv_elements * kv_dtype.itemsize
+    if quantize_kv_cache:
+        # One scalar scale per key and per value tensor, stored at ``dtype``.
+        total_bytes += model_config.nlayers * 2 * dtype.itemsize
+
+    # Classification (binary or multiclass) runs the ManyClassDecoder, which
+    # attends to the cached train activations as retrieval keys; regression
+    # (max_num_classes == 0) does not. classification == "has any classes".
+    # These activations are not quantized with the KV cache -> sized at dtype.
+    is_classification = model_config.max_num_classes > 0
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: another advantage for passing the task type would be that we don't need this logic here.

In fact, in the model we have
`task_type = "multiclass" if config.max_num_classes >= 2 else "regression"`.

would be dangerous if this is different here.

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 13]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+        inducing_elements = (
+            model_config.dist_embed_num_blocks
+            * n_features
+            * model_config.dist_embed_num_inducing_points
+            * model_config.embed_dim
+        )
+        total_bytes += inducing_elements * dtype.itemsize
+
+    # Fitted scaler stats: mean + std, each (n_features,). Always cached.
+    total_bytes += 2 * n_features * dtype.itemsize
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: nit: super minor point but are you sure the scaler stats are stored in 16-bit?

File: src/tabpfn/architectures/tabpfn_v3.py [hunk 14]
```diff
@@ -251,6 +256,133 @@ def quantize(self, dtype: torch.dtype = torch.int8) -> TabPFNV3Cache:
...
+            ``inducing_hidden`` term (see point 3).
+
+    Returns:
+        Cache size in bytes. Divide by ``1024 ** 2`` for MB.
+    """
+    # Accept either an array-like (read ``.shape``) or a (n_rows, n_features) tuple.
+    if hasattr(X_train, "shape"):
+        n_train, n_features = int(X_train.shape[0]), int(X_train.shape[1])
+    else:
+        n_train, n_features = (int(v) for v in X_train)
```
Comment (bejaeger) [inline src/tabpfn/architectures/tabpfn_v3.py]: nit: should we just pass "num_train_rows" and "num_features" instead?

**Review by bejaeger — CHANGES_REQUESTED** [review]
Thanks a lot @jmkuebler ! I left a few comments on the design.

I was wondering if we should also hook this up to the main classifier/regressor interface? Maybe via the inference engines in `inference.py`. If a wrapper doesn't support a `kv_cache` it could return None. WDYT?

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 8 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: I updated to just calcluate the full cache always, as we have no foreseeable use in mind to leave parts away. As of writing, this works for regression and classification, so the calculation does not distinguish. If in the future we reduce the cache for regression, we have to adjust the calculation accordingly.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 9 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: grew a bit organically (I first started with KV, then added the otehrs). I now changed the interface to always sum up all the terms, and removed those arguments for simplicity.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 12 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: removed now entirely, as for the time being the calculation does not seperate classiication and regression

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 14 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: had the discussion w/ Georg above. I went for the overloading such that both work. But fine to move exlusively to shapes if you prefer that.

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 13 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: It is actually not such a minor thing unfortunately. The dtype of the scale seems to depend on which mode the model runs in. If it is forced 16 bit, things might be different than when it is autocast or on CPU.

I need to take a deeper look here and also test on a GPU to make sure I get this right..

Putting the PR on draft so we don't accidentally merge until that is resolved.

Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: Ok fixed, I now handle things seperately depending whether we are on CPU with forced dtype or whether it is GPU using autocast. 🥳

Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: also tested on GPU ✅

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 10 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py:265]: >Maybe also change dtype to base_dtype in that case? 

✅
 
>should we make this more explicit and have kv_cache_dtype directly? 

Let's keep it as is for now. We only have one format for KV, but that implicitly also determines that scales are per-tensor. The exact format thus is not solely determined by the KV type, and if we add more flexibility there, we have to properly implement it including the scales.

**Comment by jmkuebler** [comment]
@bejaeger thanks a ton for your comments. I addressed most comments, but left two to resolve for you.
In particular I now added the exact dtype support both for CPU and GPU. The autocast creates different precision in the cache, which we now handle correctly and test for.

ready from my side now.

**Comment by jmkuebler** [comment]
@bejaeger 
>I was wondering if we should also hook this up to the main classifier/regressor interface? Maybe via the inference engines in inference.py. If a wrapper doesn't support a kv_cache it could return None. WDYT?

I am currently not sure about the pros and cons of this one. We could also land this PR and I take the addition to the other interfaces as follow-up and we discuss this briefly?

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 14 above)
Comment (bejaeger) [reply src/tabpfn/architectures/tabpfn_v3.py]: hm...this adds unnecessary complexity IMO so I'd prefer the simpler interface instead. But not a strict blocker from my side, since this was also discussed with @ggprior .

**Review by bejaeger — APPROVED** [review]
Looks great! thanks for all the changes.

I agree that creating a proper API can be done in a follow-up.
I think we should also add this `fn` to the other model versions and allow getting sth like `get_cache_size("v2.5", ...)` similar to `get_architecture` in `model_loading.py`

File: src/tabpfn/architectures/tabpfn_v3.py (hunk 14 above)
Comment (jmkuebler) [reply src/tabpfn/architectures/tabpfn_v3.py]: Actually I jsut did it. If the need to pass the full data comes up again, we can still ad it.
